--- ../../t_toolkit.h	2008-11-12 17:12:04.746069197 +0100
+++ t_toolkit.h	2008-11-12 17:45:54.551826912 +0100
@@ -210,10 +210,14 @@
 xdirname(const char *restrict path)
 {
     char *dirn;
+    char *garbage;
 
     assert_not_null(path);
 
-    if ((dirn = dirname(path)) == NULL)
+    dirn = dirname(garbage = xstrdup(path));
+    free(garbage);  /* no more needed */
+
+    if (dirn == NULL)
         err(errno, NULL);
 
     return (xstrdup(dirn));
